java-connectors/kafka-connect-azure-eventhubs/src/main/java/io/lenses/java/streamreactor/connect/azure/eventhubs/source/EventHubsKafkaConsumerController.java

```diff
@@ -0,0 +1,144 @@
+    consumerProperties.put(SaslConfigs.SASL_MECHANISM,
+        azureEventHubsConfig.getString(
+            getPrefixedKafkaConsumerConfigKey(SaslConfigs.SASL_MECHANISM)));
+    consumerProperties.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG,
+        azureEventHubsConfig.getString(
+            getPrefixedKafkaConsumerConfigKey(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG)));
+    consumerProperties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+        azureEventHubsConfig.getClass(
+            getPrefixedKafkaConsumerConfigKey(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG)));
+    consumerProperties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+        azureEventHubsConfig.getClass(
+            getPrefixedKafkaConsumerConfigKey(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG)));
+    consumerProperties.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+
+    return consumerProperties;
+  }
+
+  /**
+   * This method leverages BlockingQueue mechanism that BlockingQueuedKafkaConsumer puts EventHub records
+   * into. It tries to poll the queue then returns list of SourceRecords
+   *
+   * @param duration how often to poll.
+   * @return list of SourceRecords (can be empty if it couldn't poll from queue)
+   * @throws InterruptedException if interrupted while polling
+   */
+  public List<SourceRecord> poll(Duration duration) throws InterruptedException {
+    boolean includeHeaders = azureEventHubsConfig.getBoolean(
+        AzureEventHubsConfigConstants.INCLUDE_HEADERS);
+    List<SourceRecord> sourceRecords = null;
+
+    queuedKafkaConsumer.startPolling(duration);
+
+    ConsumerRecords<String, String> consumerRecords = null;
+    try {
+      consumerRecords = recordsQueue.poll(
+          duration.get(ChronoUnit.SECONDS), TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      log.info("{} has been interrupted on poll", this.getClass().getSimpleName());
+      throw e;
+    }
+
+    if (consumerRecords != null && !consumerRecords.isEmpty()) {
+      sourceRecords = new ArrayList<>(consumerRecords.count());
+      for (ConsumerRecord<String, String> consumerRecord : consumerRecords) {
+
+        AzureTopicPartitionKey azureTopicPartitionKey = new AzureTopicPartitionKey(
+            consumerRecord.topic(), consumerRecord.partition());
+        AzureOffsetMarker offsetMarker = new AzureOffsetMarker(consumerRecord.offset());
+
+        SourceRecord sourceRecord = includeHeaders
+            ? mapSourceRecordIncludingHeaders(consumerRecord, azureTopicPartitionKey,
+            offsetMarker, Schema.OPTIONAL_STRING_SCHEMA, Schema.STRING_SCHEMA) :
+            mapSourceRecordWithoutHeaders(consumerRecord, azureTopicPartitionKey,
+                offsetMarker, Schema.OPTIONAL_STRING_SCHEMA, Schema.STRING_SCHEMA);
+
+        sourceRecords.add(sourceRecord);
+      }
+    }
+    return sourceRecords != null ? sourceRecords : Collections.emptyList();
+  }
+
+  void subscribeToTopic(String topic) {
```

Comment:
I think this indirection does not help. I am repeating myself here, but the consumer and its subscription can be created at once since everything is driven by config which does not change for the lifspam of a connector task. 

In this case the task calls the controller to subscribe, the controller calls the blockingconsumer to subscribe. Testing the code becomes tedious.

Reply:
Yeah, I addressed that